New extension: Last Year Today

qcrao opened the pull request.

qcrao commented:
@baibhavbista 

Thank you for the review! I’ve closed the original [MR](https://github.com/Roam-Research/roam-depot/pull/954) and opened a new one. In the updated MR, I’ve removed the logic that automatically opens the history page on load. Now, the history page will only open at the specified time. Let me know if you have any further feedback!

baibhavbista reviewed (changes requested):
Hey @qcrao sorry, it looks like I forgot to do a detailed review the last time around. 

I found some issues: 

**1. the setting for time does not seem to work** (probably because you're using diff names for reading and writing, maybe a refactor gone wrong? - "daily-update-hour" vs "hour-to-open-last-year-today-page")
<img width="673" alt="image" src="https://github.com/user-attachments/assets/4e50bb53-1b07-4308-9a0b-fd7299bfb1fd" />

**2. there are a lot of unused dependencies, could you clean those up?** (probably from init code for the extension itself)
<img width="377" alt="image" src="https://github.com/user-attachments/assets/f16a8e49-e6ba-4deb-bc8f-0e5c52f978f5" />

3. your solution to the problem of historical pages seems to be to only show that at the exact time. As you probably know yourself, issue with this is that if the user opens the graph at 9:01 am, they won't see the historical pages. 
Might I suggest a different solution? When you open the historical pages, you do something like `extensionAPI.settings.set("last-opened-day", "01-07-2025");`. And when you load the extension, you first check this setting. If value is same as today, do nothing, otherwise open the historical pages

qcrao commented:
> Hey @qcrao sorry, it looks like I forgot to do a detailed review the last time around.
> 
> I found some issues:
> 
> **1. the setting for time does not seem to work** (probably because you're using diff names for reading and writing, maybe a refactor gone wrong? - "daily-update-hour" vs "hour-to-open-last-year-today-page") <img alt="image" width="673" src="https://private-user-images.githubusercontent.com/23657255/400684785-4e50bb53-1b07-4308-9a0b-fd7299bfb1fd.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzY0ODM5MzMsIm5iZiI6MTczNjQ4MzYzMywicGF0aCI6Ii8yMzY1NzI1NS80MDA2ODQ3ODUtNGU1MGJiNTMtMWIwNy00MzA4LTlhMGItZmQ3Mjk5YmZiMWZkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAxMTAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMTEwVDA0MzM1M1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWIyNGU3MTc3N2UzNmExNWFkMTZhZjZhNjg3YTg1ZGRhMWJmOTZlZDMyNDc4MzJlMDQyM2U3MDFlMWJlODMzY2MmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.odH3u5TYmdlTeMkX8QheXLqR-vFMgNbXYKrdzgLgRXM">
> 
> **2. there are a lot of unused dependencies, could you clean those up?** (probably from init code for the extension itself) <img alt="image" width="377" src="https://private-user-images.githubusercontent.com/23657255/400685129-f16a8e49-e6ba-4deb-bc8f-0e5c52f978f5.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzY0ODM5MzMsIm5iZiI6MTczNjQ4MzYzMywicGF0aCI6Ii8yMzY1NzI1NS80MDA2ODUxMjktZjE2YThlNDktZTZiYS00ZGViLWJjOGYtMGU1YzUyZjk3OGY1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAxMTAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMTEwVDA0MzM1M1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWJhOWU4ZjA1ZjRkNzE4ZmRlM2ZjODVjYTY5NDA1Mjc4ZjNiMzZlYmVjMDRiYWEwZGQxM2Y3YzZmMWM3OWM3ZmMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.py9TbHobN5fUoFR9WVCrR7GrnwsC41kzPsFfCiQAYWk">
> 
> 3. your solution to the problem of historical pages seems to be to only show that at the exact time. As you probably know yourself, issue with this is that if the user opens the graph at 9:01 am, they won't see the historical pages.
>    Might I suggest a different solution? When you open the historical pages, you do something like `extensionAPI.settings.set("last-opened-day", "01-07-2025");`. And when you load the extension, you first check this setting. If value is same as today, do nothing, otherwise open the historical pages

@baibhavbista 

Thank you for the detailed review. Here's my response:

1. The hours setting issue has been fixed. The configuration key names have been standardized for consistency.

2. I've cleaned up the unused dependencies.

3. Thank you for the suggestion regarding the history pages implementation. I've modified the logic:
   - Users will see 3 history pages (configurable) immediately upon loading the extension
   - After initial load, pages will open based on the configured hour (defaults to 9 AM)
   - If the hour setting isn't configured, it uses the default value
   - This works well for both usage patterns:
     - Users who close Roam daily will see the history pages upon reopening
     - Users who keep Roam running will see pages at the configured time
   - Additionally, I've added keyboard shortcuts and commands for quickly opening/closing history pages

Regarding the testing issue with qcrao+last-year-today+969: I noticed that extension.js might not be updating correctly during testing. The local file works fine, but there might be caching issues in the test environment. One way to verify this is to check the default history page count in settings - it should now show 3 instead of the previous default of 1.

Could you please verify if you're experiencing similar testing issues on your end?

Please let me know if there are any other issues you'd like me to address.

<img width="928" alt="image" src="https://github.com/user-attachments/assets/9b4a570c-aa83-4c56-b44b-aaf6739ff760" />

baibhavbista commented:
@qcrao 
Thanks for letting us know about the extension.js not getting built correctly. Our best guess right now is that your update/builds happened right as we were making some changes to the build scripts, and that is why the build silently failed

I just reran the build now and it seems to be working with the latest version
<img width="739" alt="image" src="https://github.com/user-attachments/assets/bb68dcc1-8047-4b4b-9f5d-a99a63f9602c" />


btw I have a comment regarding 3
<img width="598" alt="image" src="https://github.com/user-attachments/assets/36a00543-4e81-4882-b03e-560d02e39c38" />
The usecase I am unsure of if you're considering here is "user who opens and closes their graph multiple times a day". Currently the history pages would open every time, is this intentional? My earlier suggestion of `extensionAPI.settings.set("last-opened-day", "01-07-2025");` was partly to handle this use case btw

However, this PR has been open for quite a long time without release, so I will release it. If you want to address the above comment, please do so and open up a new PR

qcrao commented:
> @qcrao Thanks for letting us know about the extension.js not getting built correctly. Our best guess right now is that your update/builds happened right as we were making some changes to the build scripts, and that is why the build silently failed
> 
> I just reran the build now and it seems to be working with the latest version <img alt="image" width="739" src="https://private-user-images.githubusercontent.com/23657255/405898876-bb68dcc1-8047-4b4b-9f5d-a99a63f9602c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzc2MTQ5OTAsIm5iZiI6MTczNzYxNDY5MCwicGF0aCI6Ii8yMzY1NzI1NS80MDU4OTg4NzYtYmI2OGRjYzEtODA0Ny00YjRiLTlmNWQtYTk5YTYzZjk2MDJjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAxMjMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMTIzVDA2NDQ1MFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWEyZmYxNzUyYTM0N2JhZWZjMmZjY2NjYTZmYmVjYjBmNGMxNDE3NTU2NjhjMWY4ZTYxMTcwMTBhYzA0M2NiMzImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.WU_VlYWgb6_tYttXkie77x4kq682m3-QnEjt_pndIQ8">
> 
> btw I have a comment regarding 3 <img alt="image" width="598" src="https://private-user-images.githubusercontent.com/23657255/405898994-36a00543-4e81-4882-b03e-560d02e39c38.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzc2MTQ5OTAsIm5iZiI6MTczNzYxNDY5MCwicGF0aCI6Ii8yMzY1NzI1NS80MDU4OTg5OTQtMzZhMDA1NDMtNGU4MS00ODgyLWIwM2UtNTYwZDAyZTM5YzM4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAxMjMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMTIzVDA2NDQ1MFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTNlNDk3MDg4ZWI5NzI0MGE0NDYwODcwM2YwZDFkMWQ4YjVkZDQ2YzNhNzA5MDI3N2Y1YTQ5ZjZkNDU1MDJmNmYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.Ik2eZO3OG5zBMwHIYweN5IQLI_2qRRhpl8xuhBpHbvA"> The usecase I am unsure of if you're considering here is "user who opens and closes their graph multiple times a day". Currently the history pages would open every time, is this intentional? My earlier suggestion of `extensionAPI.settings.set("last-opened-day", "01-07-2025");` was partly to handle this use case btw
> 
> However, this PR has been open for quite a long time without release, so I will release it. If you want to address the above comment, please do so and open up a new PR

Thank you for merging this MR!

Regarding comment 3, I believe most users don’t frequently open and close Roam multiple times a day, so this scenario might not be a common problem. For now, I’ll keep the current behavior as is. If I receive user feedback about this in the future, I’ll consider adding a toggle to limit opening the history pages to only once per day.
